src/build.rs

```diff
@@ -321,92 +321,80 @@ impl BuildQueue {
-                                    .expect("need to specify SYSROOT env var, \
-                                            or use rustup or multirust")
-                        };
-                        args.push(sys_root.to_owned());
+                }
+                if self.config.sysroot.is_empty() {
+                    let sysroot = current_sysroot()
+                                    .expect("need to specify SYSROOT env var or use rustup or multirust");
+                    args.push("--sysroot".to_owned());
+                    args.push(sysroot);
+                }
+
+                // We can't omit compilation here, because Cargo is going to expect to get
+                // dep-info for this crate, so we shell out to rustc to get that.
+                // This is not really ideal, because we are going to
+                // compute this info anyway when we run rustc ourselves, but we don't do
+                // that before we return to Cargo.
+                // FIXME Don't do this. Start our build here rather than on another thread
+                // so the dep-info is ready by the time we return from this callback.
+                // Since ProcessBuilder doesn't allow to modify args, we need to create
+                // our own command here from scratch here.
+                let rustc_exe = env::var("RUSTC").unwrap_or("rustc".to_owned());
+                let mut cmd_dep_info = Command::new(&rustc_exe);
+                for a in &args {
+                    // Emitting only dep-info is possible only for final crate type, as
+                    // as others may emit required metadata for dependent crate types
+                    if a.starts_with("--emit") && is_final_crate_type {
+                        cmd_dep_info.arg("--emit=dep-info");
+                    } else {
+                        cmd_dep_info.arg(a);
                     }
+                }
+                cmd_dep_info.envs(cmd.get_envs().iter().filter_map(|(k, v)| v.as_ref().map(|v| (k, v))));
+                if let Some(cwd) = cmd.get_cwd() {
+                    cmd_dep_info.current_dir(cwd);
+                }
 
-                    let envs = cmd.get_envs();
-                    trace!("envs: {:?}", envs);
+                cmd_dep_info.status().expect("Couldn't execute rustc");
 
-                    {
-                        let mut cmd_line_args = self.cmd_line_args.lock().unwrap();
-                        *cmd_line_args = args.clone();
-                    }
-                    {
-                        let mut cmd_line_envs = self.cmd_line_envs.lock().unwrap();
-                        *cmd_line_envs = envs.clone();
-                    }
+                // Finally store the modified cargo-generated args/envs for future rustc calls
+                args.insert(0, rustc_exe.clone());
+                *self.cmd_line_args.lock().unwrap() = args;
+                *self.cmd_line_envs.lock().unwrap() = cmd.get_envs().clone();
```

Comment:
I'm not 100% sure how rvalues work in Rust, but I assumed that when we don't bind a value, its scope is maximally limited, so here the MutexGuard would only be limited to the scope of the statement and released before the next lock. Even so, we still could theoretically, unfortunately, interleave updates here. I'll try changing that to a single Mutex, as you're suggesting